activation/poet_client_test.go

```diff
@@ -84,6 +84,43 @@ func Test_HTTPPoetClient_Submit(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func Test_HTTPPoetClient_SubmitTillCtxDeadlineReached(t *testing.T) {
+	t.Parallel()
+
+	const maxRequests = 1
+
+	ctxWithDeadline, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+	defer cancel()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /v1/submit", func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "some_error", http.StatusInternalServerError)
+	})
+
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	cfg := server.DefaultRoundConfig()
+	client, err := NewHTTPPoetClient(types.PoetServer{Address: ts.URL}, PoetConfig{
+		PhaseShift:        cfg.PhaseShift,
+		CycleGap:          cfg.CycleGap,
+		MaxRequestRetries: maxRequests,
+	}, withCustomHttpClient(ts.Client()))
+	require.NoError(t, err)
+
+	_, err = client.Submit(
+		ctxWithDeadline,
+		time.Time{},
+		nil,
+		nil,
+		types.EmptyEdSignature,
+		types.NodeID{},
+		PoetAuth{},
+	)
+
+	require.ErrorIs(t, err, context.DeadlineExceeded)
+}
```

Comment:
What we really want to check is if the client retries until the context is canceled. How about canceling the context after N tries instead? The test would be faster and it wouldn't depend on time. Consider:
```suggestion
func Test_HTTPPoetClient_SubmitTillCtxCanceled(t *testing.T) {
	t.Parallel()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	tries := 0
	mux := http.NewServeMux()
	mux.HandleFunc("POST /v1/submit", func(w http.ResponseWriter, r *http.Request) {
		tries += 1
		if tries == 3 {
			cancel()
		}
		http.Error(w, "some_error", http.StatusInternalServerError)
	})

	ts := httptest.NewServer(mux)
	defer ts.Close()

	cfg := server.DefaultRoundConfig()
	client, err := NewHTTPPoetClient(types.PoetServer{Address: ts.URL}, PoetConfig{
		PhaseShift:        cfg.PhaseShift,
		CycleGap:          cfg.CycleGap,
		MaxRequestRetries: 1,
	}, withCustomHttpClient(ts.Client()))
	require.NoError(t, err)

	_, err = client.Submit(
		ctx,
		time.Time{},
		nil,
		nil,
		types.EmptyEdSignature,
		types.NodeID{},
		PoetAuth{},
	)

	require.ErrorIs(t, err, context.Canceled)
	require.Equal(t, 3, tries)
}
```